src/SignalR/clients/cpp/src/signalrclient/transport.h

```diff
@@ -5,35 +5,29 @@
 
 #include "pplx/pplxtasks.h"
 #include "signalrclient/transport_type.h"
+#include "transfer_format.h"
 #include "logger.h"
 
 namespace signalr
 {
     class transport
     {
     public:
-        virtual pplx::task<void> connect(const std::string &url) = 0;
-
-        virtual pplx::task<void> send(const std::string &data) = 0;
-
-        virtual pplx::task<void> disconnect() = 0;
-
         virtual transport_type get_transport_type() const = 0;
 
         virtual ~transport();
 
-    protected:
-        transport(const logger& logger, const std::function<void(const std::string &)>& process_response_callback,
-            std::function<void(const std::exception&)> error_callback);
+        virtual void start(const std::string& url, transfer_format format, std::function<void(std::exception_ptr)> callback) = 0;
+        virtual void stop(std::function<void(std::exception_ptr)> callback) = 0;
+        virtual void on_close(std::function<void(std::exception_ptr)> callback) = 0;
 
-        void process_response(const std::string &message);
-        void error(const std::exception &e);
+        virtual void send(std::string payload, std::function<void(std::exception_ptr)> callback) = 0;
 
-        logger m_logger;
+        virtual void on_receive(std::function<void(std::string, std::exception_ptr)> callback) = 0;
```

Comment:
I'd expect to use the same API in `websocket_client` as we do here, so maybe we should change `websocket_client` to use `on_receive` as well...

Reply:
The question is, do we expect people who provide their own websocket impl to do the receive loop and call our callback, or do we do the receive loop in the transport and just call receive on their impl?